apps/avifenc.c

```diff
@@ -398,6 +399,68 @@ static avifBool avifImageSplitGrid(const avifImage * gridSplitImage, uint32_t gr
     return AVIF_TRUE;
 }
 
+static int countLeadingZeros(uint32_t n)
+{
+    assert(n != 0);
+    const uint32_t msb = 1u << 31;
+    int count = 0;
+    while ((n & msb) == 0) {
+        ++count;
+        n <<= 1;
+    }
+    return count;
+}
+
+static int floorLog2(uint32_t n)
+{
+    assert(n > 0);
+    return 31 ^ countLeadingZeros(n);
+}
+
+// Set the tile configuration: the number of tiles and the tile size.
+//
+// Tiles improve encoding and decoding speeds when multiple threads are
+// available. However, for image coding, the total tile boundary length affects
+// the compression efficiency because intra prediction can't go across tile
+// boundaries. So the more tiles there are in an image, the worse the
+// compression ratio is. For a given number of tiles, making the tile size close
+// to a square tends to reduce the total tile boundary length inside the image.
+// Use more tiles along the longer dimension of the image to make the tile size
+// closer to a square.
+static void setTileConfiguration(int threads, uint32_t width, uint32_t height, int * tileRowsLog2, int * tileColsLog2)
+{
+    *tileRowsLog2 = 0;
+    *tileColsLog2 = 0;
+    if (threads > 1) {
+        // Avoid small tiles because they are particularly bad for image coding.
+        //
+        // Use no more tiles than the number of threads. Aim for one tile per
+        // thread. Using more than one thread inside one tile could be less
+        // efficient. Using more tiles than the number of threads would result in a
+        // compression penalty without much benefit.
+        const uint32_t kMinTileArea = 512 * 512;
+        const uint32_t kMaxTiles = 8;
+        uint32_t imageArea = width * height;
+        uint32_t tiles = (imageArea + kMinTileArea - 1) / kMinTileArea;
+        if (tiles > kMaxTiles) {
+            tiles = kMaxTiles;
+        }
+        if (tiles > (uint32_t)threads) {
+            tiles = threads;
+        }
+        int tilesLog2 = floorLog2(tiles);
+        // If the image's width is greater than the height, use more tile columns
+        // than tile rows.
+        if (width >= height) {
+            *tileRowsLog2 = tilesLog2 / 2;
+            *tileColsLog2 = tilesLog2 - *tileRowsLog2;
```

Comment:
Consider this extreme example:
- 16384x64px so 4 tiles
- tilesLog2 = 2, tileRowsLog2 = 1, tileColsLog2 = 1

I would imagine splitting 16384x64px into 2x2 tiles is worse than 4x1 tiles.
The following or a similar ratio might help:
```cpp
*tileRowsLog2 = (tilesLog2 * height + (((width + height) >> 1)) / (width + height);
```

Unit testing this feature would be reassuring, unfortunately there is no way to get the number of tiles at decoding currently.

Reply:
Done. Thanks for the suggestion.

It should suffice to unit test the `setTileConfiguration()` function. That will require moving `setTileConfiguration()` to some other file, such as apps/shared/avifutil.{h,c}.